docs/design/PT-166642114-miner-activated-consensus.md

```diff
@@ -0,0 +1,158 @@
+computing the outcome of the miner signalling
+and storing such outcome (true / false) in an auxiliary table.
+
+Callers requiring the signalling outcome for a certain block
+shall enquire such auxiliary table and, if no outcome yet:
+- For sync worker process,
+  yield (`take_a_break`).
+- For user API worker process,
+  terminate returning 503 with reason "Computing miner signalling outcome".
+- For peer connection (gossip),
+  ignore logging info message.
+- For locally mined block (hence context of `aec_conductor` process),
+  ignore logging error message (as it should not have happened).
+- For mined block submitted by Stratum server (hence context of `aec_conductor` process),
+  ignore logging error message (as it should not have happened).
+
+In order not to require the auxiliary table to be persisted,
+and in order to increase resilience of the implementation to minor software failures,
+the presence of the worker for the desired miner signalling outcome shall be checked:
+- By the caller requiring the signalling outcome for a certain block.
+- By the insertion of any key block confirming the signalling block interval before the height at which the new consensus protocol may be active (i.e. from HE to HS-1).
+
+##### Blocks not necessarily connected to genesis
+
+The current design of syncing receives blocks out of order.
+At that stage, no validation is performed on those blocks depending on the consensus protocol version.
+
+##### Blocks persisted in the database with a now-invalid consensus protocol
+
+After having run the node with miner-signalling consensus protocol upgrade configured,
+a user restarting afterwards the node with unconditional consensus protocol upgrade configured
+is not supported:
+such user is recommended to delete the storage
+and start the node with the new configuration - syncing the chain from scratch.
+Failure to do so may lead to unexpected results.
+
+#### Transactions
+
+##### Mempool
+
+###### Insertion
+
+The `aec_tx_pool` module enables inserting in the mempool only transactions valid at the consensus protocol of the top of the best valid chain.
+This requires no changes.
+
+###### Deletion
+
+When configured for miner signalling,
+after - on the best valid chain - the proposed fork height is reached and the consensus protocol version is decided,
+the node will not re-check transactions for the now-invalid protocol.
+
+After having configured the node for miner signalling,
+if afterwards the user configures a fixed consensus protocol at a certain height and restarts the node
+the node will not re-check transactions for the now-invalid protocol.
+
+##### State channel client finite state machine
+
+No structural changes needed.
+
+**TODO Handling of transactions read from database after user configuration makes consensus upgrade unconditional.**
+
+##### User API
+
+**TODO The user API returns transactions to the user. Mark those APIs as deprecated and remove those? Let user specify consensus protocol?**
```

Comment:
Re serialization of transactions: add note re that if any ambiguity on serialization of transaction depending on protocol, then the user API shall be enhanced with new optional field so that user can control the protocol version.

Reply:
Addressed in new commit.